File: app/src/main/java/de/passwordvault/view/settings/activity_about/SettingsAboutViewModel.java
package de.passwordvault.view.settings.activity_about;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import de.passwordvault.model.rest.RestCallback;
import de.passwordvault.model.rest.RestError;
import de.passwordvault.model.rest.legal.LegalRestClient;
import de.passwordvault.model.rest.legal.LocalizedLegalPage;


/**
 * Class implements the view model for the {@link SettingsAboutActivity}.
 *
 * @author  Christian-2003
 * @version 3.7.2
 */
public class SettingsAboutViewModel extends ViewModel {

    /**
     * Attribute stores the tag used with the REST client fetching the privacy policy.
     */
    public static final String TAG_PRIVACY = "privacy";

    /**
     * Attribute stores the tag used with the REST client fetching the terms of service.
     */
    public static final String TAG_TOS = "tos";


    /**
     * Attribute stores the REST client fetching data about the privacy policy.
     */
    @NonNull
    private final LegalRestClient privacyRestClient;

    /**
     * Attribute stores the REST client fetching data about the terms of service.
     */
    @NonNull
    private final LegalRestClient tosRestClient;

    /**
     * Attribute stores the REST error generated when fetching the privacy policy.
     */
    @Nullable
    private RestError privacyError;

    /**
     * Attribute stores the REST error generated when fetching the terms of service.
     */
    @Nullable
    private RestError tosError;


    /**
     * Constructor instantiates a new view model.
     */
    public SettingsAboutViewModel() {
        privacyRestClient = new LegalRestClient(TAG_PRIVACY, "privacy");
        tosRestClient = new LegalRestClient(TAG_TOS, "tos");
        privacyError = null;
        tosError = null;
    }


    /**
     * Method returns the error generated while fetching the privacy policy. This is {@code null}
     * before the REST client finished fetching data.
     *
     * @return  Error generated.
     */
    @Nullable
    public RestError getPrivacyError() {
        return privacyError;
    }

    /**
     * Method returns the error generated while fetching the terms of service. This is {@code null}
     * before the REST client finished fetching data.
     *
     * @return  Error generated.
     */
    @Nullable
    public RestError getTosError() {
        return tosError;
    }

    /**
     * Method changes the error generated while fetching the privacy policy.
     *
     * @param privacyError  New error generated while fetching.
     */
    public void setPrivacyError(@NonNull RestError privacyError) {
        this.privacyError = privacyError;
    }

    /**
     * Method changes the error generated while fetching the terms of service.
     *
     * @param tosError  New error generated while fetching.
     */
    public void setTosError(@NonNull RestError tosError) {
        this.tosError = tosError;
    }

    @Nullable
    public LocalizedLegalPage getPrivacyPage() {
        return privacyRestClient.getLegalPage();
    }

    @Nullable
    public LocalizedLegalPage getTosPage() {
        return tosRestClient.getLegalPage();
    }

    /**
     * Method begins fetching data from the REST API.
     *
     * @param callback  Callback to invoke once the REST client finishes fetching.
     */
    public void fetchData(@Nullable RestCallback callback) {
        privacyRestClient.fetch(callback);
        tosRestClient.fetch(callback);
    }

}

File: app/src/main/java/de/passwordvault/view/settings/activity_about/SettingsAboutViewModel.kt
package de.passwordvault.view.settings.activity_about

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import de.passwordvault.model.rest.RestCallback
import de.passwordvault.model.rest.RestError
import de.passwordvault.model.rest.legal.LegalRestClient
import de.passwordvault.model.rest.legal.LocalizedLegalPage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


/**
 * Class implements the view model for the screen which displays settings about the app.
 */
class SettingsAboutViewModel: ViewModel() {

    /**
     * REST client fetching info about the privacy policy.
     */
    private val privacyRestClient = LegalRestClient("privacy", "privacy")

    /**
     * REST client fetching info about the terms of service.
     */
    private val tosRestClient = LegalRestClient("tos", "tos")

    /**
     * REST error which occurred when fetching the privacy policy.
     */
    var privacyError: RestError? by mutableStateOf(null)

    /**
     * REST error which occurred when fetching the terms of service.
     */
    var tosError: RestError? by mutableStateOf(null)

    /**
     * Info about the privacy policy fetched from the REST API.
     */
    var privacyPage: LocalizedLegalPage? by mutableStateOf(null)

    /**
     * Info about the terms of service fetched from the REST API.
     */
    var tosPage: LocalizedLegalPage? by mutableStateOf(null)


    /**
     * Method initializes the view model.
     */
    fun init() {
        fetchData()
    }


    /**
     * Method fetches data from the REST API.
     */
    fun fetchData() = viewModelScope.launch(Dispatchers.IO) {
        if (!privacyRestClient.isFetching && !privacyRestClient.isFinished) {
            val privacyCallback = RestCallback { _, error ->
                privacyError = error
                if (error == RestError.SUCCESS) {
                    privacyPage = privacyRestClient.legalPage
                }
            }
            privacyRestClient.fetch(privacyCallback)
        }

        if (!tosRestClient.isFetching && !tosRestClient.isFinished) {
            val tosCallback = RestCallback { _, error ->
                tosError = error
                if (error == RestError.SUCCESS) {
                    tosPage = tosRestClient.legalPage
                }
            }
            tosRestClient.fetch(tosCallback)
        }
    }

}
